### PR TITLE
[Merged by Bors] - feat: Positivity extension for `Finset.prod`

### PR DESCRIPTION
**Pull request by YaelDillies**
Followup to #9365

From LeanAPAP

---

- [x] depends on:  #10538
- [x] depends on:  #10875

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/BigOperators/Order.lean [hunk 1]
```diff
@@ -830,3 +830,52 @@ theorem IsAbsoluteValue.map_prod [CommSemiring R] [Nontrivial R] [LinearOrderedC
...
+Note that this does not do any complicated reasoning. In particular, it does not try to feed in the
+`i ∈ s` hypothesis to local assumptions, and the only ways it can prove `s` is nonempty is if there
+is a local `s.Nonempty` hypothesis or `s = (univ : Finset α)` and `Nonempty α` can be synthesized by
+TC inference. -/
+@[positivity Finset.sum _ _]
+def evalFinsetSum : PositivityExt where eval {u α} zα pα e := do
+  match e with
+  | ~q(@Finset.sum _ $ι $instα $s $f) =>
+    let (lhs, _, (rhs : Q($α))) ← lambdaMetaTelescope f
+    let so : Option Q(Finset.Nonempty $s) ← do -- TODO: It doesn't complain if we make a typo?
```
Comment (eric-wieser) [inline Mathlib/Algebra/BigOperators/Order.lean:849]: leanprover/lean4#3126

File: Mathlib/Algebra/BigOperators/Order.lean [hunk 2]
```diff
@@ -830,3 +830,52 @@ theorem IsAbsoluteValue.map_prod [CommSemiring R] [Nontrivial R] [LinearOrderedC
...
+      try
+        let _fi ← synthInstanceQ q(Fintype $ι)
```
Comment (eric-wieser) [inline Mathlib/Algebra/BigOperators/Order.lean:851]: This is unused

File: Mathlib/Algebra/BigOperators/Order.lean [hunk 3]
```diff
@@ -830,3 +830,52 @@ theorem IsAbsoluteValue.map_prod [CommSemiring R] [Nontrivial R] [LinearOrderedC
...
+        let _no ← synthInstanceQ q(Nonempty $ι)
+        match s with
+        | ~q(@univ _ $fi) => pure (some q(Finset.univ_nonempty (α := $ι)))
```
Comment (eric-wieser) [inline Mathlib/Algebra/BigOperators/Order.lean:854]: This `let` should be inside the `match`.

Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Order.lean:854]: I moved it, but I'm not sure why you say "should". Are both spellings not functionally equivalent?

File: Mathlib/Algebra/BigOperators/Order.lean (hunk 1 above)
Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Order.lean]: Referenced

**Comment by YaelDillies** [comment]
`Computability.AkraBazzi` errors with
```
application type mismatch
  sum_nonneg fun i_1 x ↦
    div_nonneg
      (le_trans (Mathlib.Meta.Positivity.nonneg_of_isNat (Mathlib.Meta.NormNum.isNat_ofNat ℝ Nat.cast_zero)) g_pos)
      (rpow_nonneg_of_nonneg (Nat.cast_nonneg n) (p a b + 1))
argument has type
  ∀ i_1 ∈ range (r i n), 0 ≤ g ↑n / ↑n ^ (p a b + 1)
but function has type
  (∀ i_1 ∈ range (r i n), 0 ≤ (fun u ↦ g ↑u / ↑u ^ (p a b + 1)) i_1) →
    0 ≤ ∑ i in range (r i n), (fun u ↦ g ↑u / ↑u ^ (p a b + 1)) i
```
which makes me think the function handling is wrong. However I can't seem to fix it.

File: Mathlib/Algebra/BigOperators/Order.lean (hunk 3 above)
Comment (eric-wieser) [reply Mathlib/Algebra/BigOperators/Order.lean]: > Are both spellings not functionally equivalent?

No, because the other order avoids doing a pointless typeclass search if the set was not `univ`. Probably adding an explicit annotation to `_no` fixes the error.

Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Order.lean]: What do you mean by "adding an explicit annotation to `-no`"? I've tried several things, but none worked.

Comment (eric-wieser) [reply Mathlib/Algebra/BigOperators/Order.lean]: Oh, your problem is maybe that you can't use `return` here, it breaks Qq. Try pure instead?

Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Order.lean]: The following fails still:
```
        match s with
        | ~q(@univ _ $fi) => do
          let no : Q(Nonempty $ι) ← synthInstanceQ q(Nonempty $ι)
          pure (some q(@Finset.univ_nonempty $ι $fi $no))
```

**Comment by eric-wieser** [comment]
Here's a simpler test-case:
```lean
example (s : Finset (ℕ)) : 0 ≤ ∑ j in s, j := by positivity  --ok
example (s : Finset (ℕ)) : 0 ≤ s.sum id := by positivity -- fails
```

**Comment by alexjbest** [comment]
I pushed an attempt at a fix, we beta expand if the f is not already a lambda

File: Mathlib/Algebra/BigOperators/Order.lean [hunk 4]
```diff
@@ -845,7 +845,7 @@ TC inference. -/
 def evalFinsetSum : PositivityExt where eval {u α} zα pα e := do
   match e with
   | ~q(@Finset.sum _ $ι $instα $s $f) =>
-    let (lhs, _, (rhs : Q($α))) ← lambdaMetaTelescope f
+    let (lhs, _, (rhs : Q($α))) ← lambdaMetaTelescope (if f.isLambda then f else q(fun x => $f x))
```
Comment (eric-wieser) [inline Mathlib/Algebra/BigOperators/Order.lean:848]: This looks risky if you have `f := fun {ohno} (x) => ohno`; though I suspect this is not the right approach anyway.

Comment (alexjbest) [reply Mathlib/Algebra/BigOperators/Order.lean:848]: Sorry I'm not sure I understand your example can you elaborate on that the issue would be?

Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Order.lean:848]: `f := fun {ohno} (x) => ohno` can't happen as an argument to `Finset.sum`?

Comment (eric-wieser) [reply Mathlib/Algebra/BigOperators/Order.lean:848]: > can you elaborate on that the issue would be?

`$f x` assumes that the argument to `f` is explicit



> `f := fun {ohno} (x) => ohno` can't happen as an argument to `Finset.sum`?

Isn't `Finset.sum univ fun {ohno : ℕ} (x : ℕ) => ohno` legal?

Comment (alexjbest) [reply Mathlib/Algebra/BigOperators/Order.lean:848]: Oh sure, but in your example `(fun {ohno : ℕ} (x : ℕ) => ohno) 1` is just id in function space, which seems like it would work fine.
If you can come up with a realistic mw `example` I'll try and fix, but I can't really think of any myself.

Comment (eric-wieser) [reply Mathlib/Algebra/BigOperators/Order.lean:848]: I don't know what you mean by it being `id`; it's not a function at all  `(fun {ohno : ℕ} (x : ℕ) => ohno) 1 : ℕ`, and it reduces to `?ohno : ℕ` with an unresolved metavariable

Comment (eric-wieser) [reply Mathlib/Algebra/BigOperators/Order.lean:848]: `Finset.sum univ (id fun {n : ℕ} => n)` feels like a more realistic test

Comment (alexjbest) [reply Mathlib/Algebra/BigOperators/Order.lean:848]: I see, I wasn't aware it would eagerly apply implicits in this way

Comment (alexjbest) [reply Mathlib/Algebra/BigOperators/Order.lean:848]: I mean like realistic in the sense of something someone will be applying positivity too. Its nice to be feature complete but I'm not sure how much time I want to spend fixing things that will never come up in practice, at the moment its still not clear to me whether the issue you're pointing out is a real issue for this tactic or not

Comment (eric-wieser) [reply Mathlib/Algebra/BigOperators/Order.lean:848]: My example was more to demonstrate that this approach seems somewhat fishy; it's fine for positivity to not support these cases, but it should never produce invalid proof terms.

File: Mathlib/Algebra/BigOperators/Order.lean [hunk 5]
```diff
@@ -816,3 +816,55 @@ theorem IsAbsoluteValue.map_prod [CommSemiring R] [Nontrivial R] [LinearOrderedC
...
+      let pα' ← synthInstanceQ q(OrderedAddCommMonoid $α)
+      assertInstancesCommute
+      let pr : Q(∀ i, 0 ≤ $f i) ← mkLambdaFVars #[i] pb
+      return .nonnegative q(@sum_nonneg $ι $α $pα' $f $s fun i _ ↦ $pr i)
+    | .positive pb, .some (fi : Q(Finset.Nonempty $s)) => do
+      let pα' ← synthInstanceQ q(OrderedCancelAddCommMonoid $α)
+      assertInstancesCommute
+      let pr : Q(∀ i, 0 < $f i) ← mkLambdaFVars #[i] pb
+      return .positive q(@sum_pos $ι $α $pα' $f $s (fun i _ ↦ $pr i) $fi)
+    | _, _ => pure .none
```
Comment (eric-wieser) [inline Mathlib/Algebra/BigOperators/Order.lean:857]: What if `body` is positive but `so` is `none`; don't we still want to prove that the sum is non-negative?

Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Order.lean]: Fixed. It makes the extension longer but hopefully more robust.

File: Mathlib/Algebra/BigOperators/Order.lean [hunk 6]
```diff
@@ -816,3 +816,65 @@ theorem IsAbsoluteValue.map_prod [CommSemiring R] [Nontrivial R] [LinearOrderedC
...
+        try
+          match s with
+          | ~q(@univ _ $fi) => do
+            let _no ← synthInstanceQ q(Nonempty $ι)
+            return some q(Finset.univ_nonempty (α := $ι))
+          | _ => throwError "`s` is not `univ`"
+        catch _ => do
+          let .some fv ← findLocalDeclWithType? q(Finset.Nonempty $s) | pure none
+          pure (some (.fvar fv))
+        | failure -- Fail if the body is not nonempty
```
Comment (eric-wieser) [inline Mathlib/Algebra/BigOperators/Order.lean:853]: Isn't the failure unreachable here? You never return `non` from the `do` block

File: Mathlib/Algebra/BigOperators/Order.lean [hunk 7]
```diff
@@ -816,3 +825,147 @@ theorem IsAbsoluteValue.map_prod [CommSemiring R] [Nontrivial R] [LinearOrderedC
...
+        try
+          match s with
+          | ~q(@univ _ $fi) => do
+            let _no ← synthInstanceQ q(Nonempty $ι)
+            pure q(Finset.univ_nonempty (α := $ι))
+          | _ => throwError "`s` is not `univ`"
+        catch _ => do
+          let .some fv ← findLocalDeclWithType? q(Finset.Nonempty $s)
+            | failure -- Fail if the set is not provably nonempty
+          pure (.fvar fv)
```
Comment (Vierkantor) [inline Mathlib/Algebra/BigOperators/Order.lean:869]: I feel like this should be its own definition, since I can imagine there are lots more usecases for checking that a finset is not empty. Maybe even an `aesop` ruleset?

File: Mathlib/Algebra/BigOperators/Order.lean [hunk 8]
```diff
@@ -816,3 +825,147 @@ theorem IsAbsoluteValue.map_prod [CommSemiring R] [Nontrivial R] [LinearOrderedC
...
+      -- then the type-ascription is ignored. See leanprover/lean4#3126
+      let (ps : Q(Finset.Nonempty $s)) ← do
+        try
+          match s with
+          | ~q(@univ _ $fi) => do
+            let _no ← synthInstanceQ q(Nonempty $ι)
+            pure q(Finset.univ_nonempty (α := $ι))
+          | _ => throwError "`s` is not `univ`"
+        catch _ => do
+          let .some fv ← findLocalDeclWithType? q(Finset.Nonempty $s)
```
Comment (eric-wieser) [inline Mathlib/Algebra/BigOperators/Order.lean:867]: Should we do these in the other order? I'd argue that finding the assumption is faster than a typeclass search, and so we should try it first.

File: Mathlib/Analysis/NormedSpace/BoundedLinearMaps.lean [hunk 9]
```diff
@@ -215,24 +215,18 @@ operation. -/
...
-        constructor
-        · exact (p.1.le_opNorm m).trans (mul_le_mul_of_nonneg_right (norm_fst_le p)
-            (Finset.prod_nonneg fun i _ => norm_nonneg _))
-        · exact (p.2.le_opNorm m).trans (mul_le_mul_of_nonneg_right (norm_snd_le p)
-            (Finset.prod_nonneg fun i _ => norm_nonneg _))⟩ }
+      p.1.prod p.2 where
+  map_add p₁ p₂ := by ext : 1; rfl
+  map_smul c p := by ext : 1; rfl
+  bound := by
+    refine ⟨1, zero_lt_one, fun p ↦ ?_⟩
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Analysis/NormedSpace/BoundedLinearMaps.lean:222]: This part of the diff belongs in #10538

File: Mathlib/Algebra/BigOperators/Order.lean [hunk 10]
```diff
@@ -680,8 +660,37 @@ theorem prod_lt_prod_of_nonempty (hf : ∀ i ∈ s, 0 < f i) (hfg : ∀ i ∈ s,
   obtain ⟨i, hi⟩ := h_ne
   exact ⟨i, hi, hfg i hi⟩
 
+end PosMulStrictMono
+end CommMonoidWithZero
+
+section StrictOrderedCommSemiring
+
 end StrictOrderedCommSemiring
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Order.lean:668]: This part of the diff belongs in #10538

File: Mathlib/Algebra/BigOperators/Order.lean [hunk 11]
```diff
@@ -816,3 +825,147 @@ theorem IsAbsoluteValue.map_prod [CommSemiring R] [Nontrivial R] [LinearOrderedC
...
+  have : Finset.Nonempty {1} := singleton_nonempty 1
+  positivity
+example (s : Finset ℕ) : 0 ≤ ∏ j in s, j := by positivity
+example (s : Finset ℕ) : 0 ≤ s.sum id := by positivity
+example (s : Finset ℕ) (f : ℕ → ℕ) (a : ℕ) : 0 ≤ s.sum (f a) := by positivity
+
+-- Make sure that the extension doesn't produce an invalid term by accidentally unifying `?n` with
+-- `0` because of the `hf` assumption
+set_option linter.unusedVariables false in
+example (f : ℕ → ℕ) (hf : 0 ≤ f 0) : 0 ≤ ∏ n in Finset.range 10, f n := by positivity
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Order.lean:969]: Now in #10538

File: Mathlib/Algebra/BigOperators/Order.lean (hunk 7 above)
Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Order.lean:869]: I don't really know how to set up an aesop ruleset, so it's a TODO for now.

File: Mathlib/Algebra/BigOperators/Order.lean
Comment (eric-wieser) [inline Mathlib/Algebra/BigOperators/Order.lean:1]: Can you split all these generalizations to their own PR? It's best not to mix refactors and meta code.

File: Mathlib/Algebra/BigOperators/Order.lean [hunk 12]
```diff
@@ -859,4 +868,56 @@ def evalFinsetSum : PositivityExt where eval {u α} zα pα e := do
...
+    try
+      let .positive pbody := rbody | failure -- Fail if the body is not provably positive
+      let instαmon ← synthInstanceQ q(CommMonoidWithZero $α)
+      let instαzeroone ← synthInstanceQ q(ZeroLEOneClass $α)
+      let instαposmul ← synthInstanceQ q(PosMulStrictMono $α)
+      let instαnontriv ← synthInstanceQ q(Nontrivial $α)
+      assertInstancesCommute
+      let pr : Q(∀ i, 0 < $f i) ← mkLambdaFVars #[i] pbody
+      pure $ .positive q(@prod_pos $ι $α $instαmon $pα $instαzeroone $instαposmul $instαnontriv $f
+        $s fun i _ ↦ $pr i)
```
Comment (eric-wieser) [inline Mathlib/Algebra/BigOperators/Order.lean:901]: why not:
```suggestion
      pure $ .positive q(prod_pos fun i _ ↦ $pr i)
```

File: Mathlib/Algebra/BigOperators/Order.lean
Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Order.lean:1]: #10875

File: Mathlib/Algebra/BigOperators/Order.lean (hunk 12 above)
Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Order.lean]: Because there is a bug in `assertInstancesCommute` that makes it loop when you have several unnamed assumptions in your local context. Added a TODO but we should open an issue.

File: Mathlib/Algebra/BigOperators/Order.lean [hunk 13]
```diff
@@ -867,4 +867,54 @@ def evalFinsetSum : PositivityExt where eval {u α} zα pα e := do
...
+      assertInstancesCommute
+      pure $ .nonnegative q(prod_nonneg fun i _ ↦ $pr i)
+    catch _ =>
+      let pbody ← rbody.toNonzero
+      let pr : Q(∀ i, $f i ≠ 0) ← mkLambdaFVars #[i] pbody
+      -- TODO: We must name the following, else `assertInstancesCommute` loops.
+      let _instαmon ← synthInstanceQ q(CommMonoidWithZero $α)
+      let _instαnontriv ← synthInstanceQ q(Nontrivial $α)
+      let _instαnozerodiv ← synthInstanceQ q(NoZeroDivisors $α)
+      assertInstancesCommute
```
Comment (adomani) [inline Mathlib/Algebra/BigOperators/Order.lean:916]:
```suggestion
    let rbody ← core zα pα body
    -- TODO: We must name the following, else `assertInstancesCommute` loops.
    let _instαmon ← synthInstanceQ q(CommMonoidWithZero $α)
    let _instαzeroone ← synthInstanceQ q(ZeroLEOneClass $α)
    let _instαposmul ← synthInstanceQ q(PosMulStrictMono $α)
    let _instαnontriv ← synthInstanceQ q(Nontrivial $α)
    let _instαnozerodiv ← synthInstanceQ q(NoZeroDivisors $α)
    assertInstancesCommute
    -- Try to show that the sum is positive
    try
      let .positive pbody := rbody | failure -- Fail if the body is not provably positive
      let pr : Q(∀ i, 0 < $f i) ← mkLambdaFVars #[i] pbody
      pure $ .positive q(prod_pos fun i _ ↦ $pr i)
    -- Try to show that the sum is nonnegative
    catch _ => try
      let pbody ← rbody.toNonneg
      let pr : Q(∀ i, 0 ≤ $f i) ← mkLambdaFVars #[i] pbody
      pure $ .nonnegative q(prod_nonneg fun i _ ↦ $pr i)
    catch _ =>
      let pbody ← rbody.toNonzero
      let pr : Q(∀ i, $f i ≠ 0) ← mkLambdaFVars #[i] pbody
      pure $ .nonzero q(prod_ne_zero fun i _ ↦ $pr i)
```
moving to earlier and deduplicating makes the tests still work locally: is this a reasonable change?

What I can see that might fail is if you some of those `synthInstanceQ` commands fails only in the `catch`, but then maybe there should be a test against that.

Note: I have no experience of working with `Qq`, so feel free to give profuse comments about why this is a bad idea for my own edification!

Comment (adomani) [reply Mathlib/Algebra/BigOperators/Order.lean:916]: Update: the suggested changes pass CI.

https://github.com/leanprover-community/mathlib4/actions/runs/8168589642/job/22330941550

Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Order.lean:916]: > What I can see that might fail is if some of those `synthInstanceQ` commands fails only in the `catch`, but then maybe there should be a test against that.

I'm not sure this is what you meant, but indeed your version has a problem if the body is strictly positive but multiplication isn't strictly monotone. I have now added a test that works with my version but not yours.

File: Mathlib/Algebra/BigOperators/Order.lean [hunk 14]
```diff
@@ -867,4 +867,56 @@ def evalFinsetSum : PositivityExt where eval {u α} zα pα e := do
...
+def evalFinsetProd : PositivityExt where eval {u α} zα pα e := do
+  match e with
+  | ~q(@Finset.prod _ $ι $instα $s $f) =>
+    let i : Q($ι) ← mkFreshExprMVarQ q($ι) .syntheticOpaque
+    have body : Q($α) := Expr.betaRev f #[i]
+    let rbody ← core zα pα body
+    let _instαmon ← synthInstanceQ q(CommMonoidWithZero $α)
+    let p_pos : Option Q(0 < $e) := ← (do
+      let .positive pbody := rbody | pure none -- Fail if the body is not provably positive
+      -- TODO: We must name the following, else `assertInstancesCommute` loops.
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Order.lean:891]: I haven't yet decided. Please give your input here: https://github.com/leanprover-community/mathlib4/pull/9365#discussion_r1500378763

File: Mathlib/Algebra/BigOperators/Order.lean (hunk 12 above)
Comment (dwrensha) [reply Mathlib/Algebra/BigOperators/Order.lean]: I opened https://github.com/leanprover-community/quote4/issues/38.

**Review by thorimur — COMMENTED** [review]
Love how much can be replaced with `by positivity` due to this PR! :)

This review is shorter than it appears: almost all of these 12 comments are about `$`, with only a couple potentially "meaningful" review comments.

File: Mathlib/Algebra/BigOperators/Order.lean (hunk 13 above)
Comment (adomani) [reply Mathlib/Algebra/BigOperators/Order.lean]: Ok, this is exactly what I meant, thanks!

File: Mathlib/Algebra/BigOperators/Order.lean [hunk 15]
```diff
@@ -867,4 +867,56 @@ def evalFinsetSum : PositivityExt where eval {u α} zα pα e := do
...
+    if let some p_pos := p_pos then return .positive p_pos
+    let p_nonneg : Option Q(0 ≤ $e) := ← (do
+      let .some pbody := rbody.toNonneg
+        | return none -- Fail if the body is not provably nonnegative
+      let pr : Q(∀ i, 0 ≤ $f i) ← mkLambdaFVars #[i] pbody
+      -- TODO: We must name the following, else `assertInstancesCommute` loops.
+      let .some _instαzeroone ← trySynthInstanceQ q(ZeroLEOneClass $α) | pure none
+      let .some _instαposmul ← trySynthInstanceQ q(PosMulMono $α) | pure none
+      assertInstancesCommute
+      return some q(prod_nonneg fun i _ ↦ $pr i))
```
[the remark this answers is not in the document]
Comment (Vierkantor) [reply Mathlib/Algebra/BigOperators/Order.lean:908]: Maybe this whole block can be moved to a separate definition? It's a long definition anyway.

**Review by Vierkantor — COMMENTED** [review]
Looks good with @thorimur's comments. (I am neutral about the `<|` vs `$` issue, but the style guide takes precedence.) @robertylewis, any suggestions or shall we delegate it to @thorimur?

File: Mathlib/Algebra/BigOperators/Order.lean (hunk 15 above)
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Order.lean:908]: Sorry what do you mean, @Vierkantor ? It's two separate blocks

File: Mathlib/Algebra/BigOperators/Order.lean [hunk 16]
```diff
@@ -867,4 +867,56 @@ def evalFinsetSum : PositivityExt where eval {u α} zα pα e := do
...
+    let _instαmon ← synthInstanceQ q(CommMonoidWithZero $α)
+    let p_pos : Option Q(0 < $e) := ← (do
+      let .positive pbody := rbody | pure none -- Fail if the body is not provably positive
+      -- TODO: We must name the following, else `assertInstancesCommute` loops.
+      let .some _instαzeroone ← trySynthInstanceQ q(ZeroLEOneClass $α) | pure none
+      let .some _instαposmul ← trySynthInstanceQ q(PosMulStrictMono $α) | pure none
+      let .some _instαnontriv ← trySynthInstanceQ q(Nontrivial $α) | pure none
+      assertInstancesCommute
+      let pr : Q(∀ i, 0 < $f i) ← mkLambdaFVars #[i] pbody
+      return some q(prod_pos fun i _ ↦ $pr i))
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Order.lean:897]: My lambda handling skills are very shaky, so I would rather not risk editing it. Regardless, I doubt you could catch such an inconsequential difference in a benchmark

File: Mathlib/Algebra/BigOperators/Order.lean [hunk 17]
```diff
@@ -867,4 +867,56 @@ def evalFinsetSum : PositivityExt where eval {u α} zα pα e := do
...
+    let rbody ← core zα pα body
+    let _instαmon ← synthInstanceQ q(CommMonoidWithZero $α)
+    let p_pos : Option Q(0 < $e) := ← (do
+      let .positive pbody := rbody | pure none -- Fail if the body is not provably positive
+      -- TODO: We must name the following, else `assertInstancesCommute` loops.
+      let .some _instαzeroone ← trySynthInstanceQ q(ZeroLEOneClass $α) | pure none
+      let .some _instαposmul ← trySynthInstanceQ q(PosMulStrictMono $α) | pure none
+      let .some _instαnontriv ← trySynthInstanceQ q(Nontrivial $α) | pure none
+      assertInstancesCommute
+      let pr : Q(∀ i, 0 < $f i) ← mkLambdaFVars #[i] pbody
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Order.lean:896]: That's indeed a strange choice of default!

File: Mathlib/Analysis/NormedSpace/Multilinear/Basic.lean [hunk 18]
```diff
@@ -1382,12 +1375,13 @@ instance completeSpace [CompleteSpace G] : CompleteSpace (ContinuousMultilinearM
   have cau : ∀ v, CauchySeq fun n => f n v := by
     intro v
     apply cauchySeq_iff_le_tendsto_0.2 ⟨fun n => b n * ∏ i, ‖v i‖, _, _, _⟩
-    · intro
-      exact mul_nonneg (b0 _) (nonneg v)
+    · intro n
+      have := b0 n
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Analysis/NormedSpace/Multilinear/Basic.lean:1379]: Kinda? This particular case is a bit scope-creepy in that some random assumption in the context needs to be fed an argument to become a positivity hypothesis. It's not the same as wanting to derive `0 ≤ ∑ i in s, f i` from an `∀ i ∈ s, 0 ≤ f i` assumption. If it were the latter, we might want a todo, but here I don't think it warrants one.

File: Mathlib/Algebra/BigOperators/Order.lean (hunk 15 above)
[the remark this answers is not in the document]
Comment (Vierkantor) [reply Mathlib/Algebra/BigOperators/Order.lean]: Sorry, I misparsed the code. I think we can format this a bit better though.

File: Mathlib/Algebra/Order/BigOperators/Ring/Finset.lean [hunk 19]
```diff
@@ -207,3 +207,61 @@ lemma IsAbsoluteValue.map_prod [CommSemiring R] [Nontrivial R] [LinearOrderedCom
 #align is_absolute_value.map_prod IsAbsoluteValue.map_prod
 
 end AbsoluteValue
+
+namespace Mathlib.Meta.Positivity
+open Qq Lean Meta Finset
+
+private alias ⟨_, prod_ne_zero⟩ := prod_ne_zero_iff
+
+/-- The `positivity` extension which proves that `∏ i in s, f i` is nonnegative if `f` is, and
```
Comment (thorimur) [inline Mathlib/Algebra/Order/BigOperators/Ring/Finset.lean:216]: (Lean 3 copypasta?)
```suggestion
/-- The `positivity` extension which proves that `∀ i, f i` is nonnegative if `f` is, and
```

Comment (YaelDillies) [reply Mathlib/Algebra/Order/BigOperators/Ring/Finset.lean:216]: Ah, certainly not! You are confusing `Π` and `∏`

Comment (thorimur) [reply Mathlib/Algebra/Order/BigOperators/Ring/Finset.lean:216]: Ah, right... 😆

**Comment by thorimur** [comment]
LGTM! :)

**Comment by eric-wieser** [comment]
!bench

**Comment by semorrison** [comment]
bors merge
